### ISSUE TITLE
Training speed improvements

### ISSUE THREAD

**Issue by ildoonet** [ildoonet/tf-pose-estimation#401]
- Data Augmentation is Slow
- Memory Consumption
- Cpu Load too high

**Comment by ildoonet** [comment]
Numba
- 1.6x-1.7x speed up
- Use less subprocesses
  - less memory consumption
  - less cpu load

TODO
- some hangups before writing logs
- some hangups before calculating validation logs

**Comment by skyler1253** [comment]
hey @ildoonet - I'm able to step through training quite a bit quicker in the dev/architecture-mobilenet2 branch (also training on mobilenet_v2).  I'm about to normalize the inputs as described in #369, but i seem to get hung up on the sess.run() calls happening every 500 steps:

sess.run([total_loss, total_loss_ll, total_loss_ll_paf, total_loss_ll_heat, learning_rate, merged_summary_op, enqueuer.size()])

These calls seem to be executing on a single CPU core and it takes a good 20 minutes.  Any ideas on how this might be sped up?

**Comment by filipetrocadoferreira** [comment]
`@staticmethod
    def put_vectormap(vectormap, countmap, plane_idx, center_from, center_to, threshold=8):
        _, height, width = vectormap.shape[:3]

        vec_x = center_to[0] - center_from[0]
        vec_y = center_to[1] - center_from[1]

        min_x = max(0, int(min(center_from[0], center_to[0]) - threshold))
        min_y = max(0, int(min(center_from[1], center_to[1]) - threshold))

        max_x = min(width, int(max(center_from[0], center_to[0]) + threshold))
        max_y = min(height, int(max(center_from[1], center_to[1]) + threshold))

        norm = math.sqrt(vec_x ** 2 + vec_y ** 2)
        if norm == 0:
            return

        vec_x /= norm
        vec_y /= norm

        x = np.arange(min_x,max_x)
        y = np.arange(min_y,max_y)

        bec_x = x - center_from[0]
        bec_y = y - center_from[1]

        dist = np.abs(bec_x * vec_y - bec_y[:,np.newaxis] * vec_x)

        countmap[plane_idx][min_y:max_y,min_x:max_x][dist<=threshold] += 1
        vectormap[plane_idx * 2 + 0][min_y:max_y,min_x:max_x][dist<=threshold] = vec_x
        vectormap[plane_idx * 2 + 1][min_y:max_y,min_x:max_x][dist<=threshold] = vec_y `

`@staticmethod
    def put_heatmap(heatmap, plane_idx, center, sigma):
        center_x, center_y = center
        _, height, width = heatmap.shape[:3]

        th = 4.6052
        delta = math.sqrt(th * 2)

        x0 = int(max(0, center_x - delta * sigma))
        y0 = int(max(0, center_y - delta * sigma))

        x1 = int(min(width, center_x + delta * sigma))
        y1 = int(min(height, center_y + delta * sigma))

        for y in range(y0, y1):
            for x in range(x0, x1):
                d = (x - center_x) ** 2 + (y - center_y) ** 2
                exp = d / 2.0 / sigma / sigma
                if exp > th:
                    continue
                heatmap[plane_idx][y][x] = max(heatmap[plane_idx][y][x], math.exp(-exp))
                heatmap[plane_idx][y][x] = min(heatmap[plane_idx][y][x], 1.0)
`


This numpy-based versions of `put_vectormap()` and `get_vectormap()` really speedup pre-processing.

**Comment by ildoonet** [comment]
@filipetrocadoferreira Thanks. I will test it. With numba, it will be faster.
